x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsProvider.java

```diff
@@ -428,18 +431,101 @@ public void timingStats(String jobId, Consumer<TimingStats> handler, Consumer<Ex
         searchSingleResult(
             jobId,
             TimingStats.TYPE.getPreferredName(),
-            createTimingStatsSearch(indexName, jobId),
+            createLatestTimingStatsSearch(indexName, jobId),
             TimingStats.PARSER,
             result -> handler.accept(result.result),
             errorHandler,
             () -> new TimingStats(jobId));
     }
 
-    private SearchRequestBuilder createTimingStatsSearch(String indexName, String jobId) {
+    private SearchRequestBuilder createLatestTimingStatsSearch(String indexName, String jobId) {
         return client.prepareSearch(indexName)
             .setSize(1)
             .setIndicesOptions(IndicesOptions.lenientExpandOpen())
-            .setQuery(QueryBuilders.idsQuery().addIds(TimingStats.documentId(jobId)));
+            .setQuery(QueryBuilders.idsQuery().addIds(TimingStats.documentId(jobId)))
+            .addSort(SortBuilders.fieldSort(TimingStats.BUCKET_COUNT.getPreferredName()).order(SortOrder.DESC));
+    }
+
+    public void datafeedTimingStats(List<String> jobIds, Consumer<Map<String, DatafeedTimingStats>> handler,
+                                    Consumer<Exception> errorHandler) {
+        if (jobIds.isEmpty()) {
+            handler.accept(new HashMap<>());
```

Comment:
It doesn't matter this time but for an empty map you could use `Collections.emptyMap()` to avoid creating a new object.

Reply:
Done. Let's save some memory ;)